### PR TITLE
Fix ignored HttpClient-decoder configuration issue

### PR DESCRIPTION
**Pull request by maximr-dev**
extracted HttpClient creation and connection-observer

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java [hunk 1]
```diff
@@ -16,128 +21,46 @@
...
-            this.httpClient = this.httpClient.wiretap(REACTOR_NETWORK_LOG_CATEGORY, LogLevel.INFO);
-        }
-
-        this.httpClient.secure(sslContextSpec -> sslContextSpec.sslContext(configs.getSslContext()))
-            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, (int) configs.getConnectionAcquireTimeout().toMillis())
-            .httpResponseDecoder(httpResponseDecoderSpec ->
-                httpResponseDecoderSpec.maxInitialLineLength(configs.getMaxHttpInitialLineLength())
-                    .maxHeaderSize(configs.getMaxHttpHeaderSize())
-                    .maxChunkSize(configs.getMaxHttpChunkSize())
-                    .validateHeaders(true));
```
Comment (maximr-dev) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:140]: The configurations were not assigned back to this.httpClient at L134, hence lost maxHttpHeaderSize configuration.

**Comment by kushagraThapar** [comment]
@maximr-dev - thanks for your contribution, I understand the bug. However, I don't think we need this big change to fix it. 

We can just fix the issue with a simple return statement here - 

```
return this.httpClient.secure(sslContextSpec -> sslContextSpec.sslContext(configs.getSslContext()))
            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, (int) configs.getConnectionAcquireTimeout().toMillis())
            .httpResponseDecoder(httpResponseDecoderSpec ->
                httpResponseDecoderSpec.maxInitialLineLength(configs.getMaxHttpInitialLineLength())
                    .maxHeaderSize(configs.getMaxHttpHeaderSize())
                    .maxChunkSize(configs.getMaxHttpChunkSize())
                    .validateHeaders(true));
```

And in the constructor, we can just re-assign the httpClient like this: 

```
reactorNettyClient.httpClient = reactorNettyClient.configureChannelPipelineHandlers();
```

With this big change (mostly refactoring), there is a large overhead of testing involved, which can delay this fix significantly. I appreciate the added unit test. 

If there is a way for you to keep the unit test with current architecture, and just fix the bug, that would be very easy to get in sooner (by next week's release). Thoughts ?

**Comment by kushagraThapar** [comment]
@maximr-dev  - or even better, we don't need to even change the method signature. 
We can just follow the same convention we have used with other configurations, like proxy and logger. 

```
this.httpClient = this.httpClient.secure(sslContextSpec -> sslContextSpec.sslContext(configs.getSslContext()))
            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, (int) configs.getConnectionAcquireTimeout().toMillis())
            .httpResponseDecoder(httpResponseDecoderSpec ->
                httpResponseDecoderSpec.maxInitialLineLength(configs.getMaxHttpInitialLineLength())
                    .maxHeaderSize(configs.getMaxHttpHeaderSize())
                    .maxChunkSize(configs.getMaxHttpChunkSize())
                    .validateHeaders(true));
                   
                   ```
That way, we don't need to change anything else. I tested it locally, and this solution works perfectly. I would incline for this, so that we don't have to refactor a bunch of classes if there is no need.

**Comment by maximr-dev** [comment]
> @maximr-dev - or even better, we don't need to even change the method signature.
> We can just follow the same convention we have used with other configurations, like proxy and logger.
> 
> ```
> this.httpClient = this.httpClient.secure(sslContextSpec -> sslContextSpec.sslContext(configs.getSslContext()))
>             .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, (int) configs.getConnectionAcquireTimeout().toMillis())
>             .httpResponseDecoder(httpResponseDecoderSpec ->
>                 httpResponseDecoderSpec.maxInitialLineLength(configs.getMaxHttpInitialLineLength())
>                     .maxHeaderSize(configs.getMaxHttpHeaderSize())
>                     .maxChunkSize(configs.getMaxHttpChunkSize())
>                     .validateHeaders(true));
>                    
> ```
> 
> That way, we don't need to change anything else. I tested it locally, and this solution works perfectly. I would incline for this, so that we don't have to refactor a bunch of classes if there is no need.

Sure, understand the scope of change was bigger, it was done mainly so that it can be unit tested, but on the other hand the fix is clear and simple, so i've limited the fix for the absolute necessity, as you've also pointed out.

**Comment by kushagraThapar** [comment]
@maximr-dev - I still believe we can add unit tests to this structure. Would you be open if I contribute to your PR with a new unit tests class. So that we can have all the changes in the same PR ?

**Comment by maximr-dev** [comment]
> @maximr-dev - I still believe we can add unit tests to this structure. Would you be open if I contribute to your PR with a new unit tests class. So that we can have all the changes in the same PR ?

Sounds good!

**Comment by kushagraThapar** [comment]
@maximr-dev  -  thanks. I have added the basic tests for reactor netty http client. Please feel free to more unit tests if you would like. Thanks.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/http/ReactorNettyHttpClientTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        this.reactorNettyHttpClient = HttpClient.createFixed(new HttpClientConfig(this.configs));
+    }
+
+    @AfterClass(groups = "unit")
+    public void after_ReactorNettyHttpClientTest() {
+        LifeCycleUtils.closeQuietly(reactorNettyHttpClient);
+    }
+
+    @Test(groups = "unit")
+    public void testHttpClientWithMaxHeaderSize() {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/http/ReactorNettyHttpClientTest.java:40]: as per our code style, we do not use "test" prefix for any test method.

please remove the 'test' prefix for this method and all other test methods in this file.

**Review by moderakh — APPROVED** [review]
LGTM. 
Thanks for contribution @maximr-dev

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/http/ReactorNettyHttpClientTest.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/http/ReactorNettyHttpClientTest.java]: @moderakh  - thanks for pointing that out. That was actually added by me. Removed the prefix now. :)

**Review by kushagraThapar — APPROVED** [review]
Thanks @maximr-dev  for your contribution and finding this bug. Looks good from bug's point of view. 
Please add any additional test cases if you would like. 
Let me know when you are finished, and I will merge the PR.

**Comment by maximr-dev** [comment]
> Thanks @maximr-dev for your contribution and finding this bug. Looks good from bug's point of view.
> Please add any additional test cases if you would like.
> Let me know when you are finished, and I will merge the PR.

Thanks for contributing the tests, I think they are sufficient for covering the case. 
I'm good with with merging the PR.

**Comment by kushagraThapar** [comment]
Intermittent timeout failures on CI, merging this PR in.

**Pull request merged**